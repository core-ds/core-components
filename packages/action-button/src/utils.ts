import { getDataTestId } from '@alfalab/core-components-shared';

export function getActionButtonTestIds(dataTestId: string) {
    return {
        actionButton: dataTestId,
        spinner: getDataTestId(dataTestId, 'loader'),
    };
}
