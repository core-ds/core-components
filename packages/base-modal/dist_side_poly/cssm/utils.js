'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var coreComponentsGlobalStore = require('@alfalab/core-components-global-store/cssm');
var coreComponentsShared = require('@alfalab/core-components-shared/cssm');

function isScrolledToTop(target) {
    return target.scrollTop <= 0;
}
function isScrolledToBottom(target) {
    return target.scrollHeight - target.offsetHeight <= target.scrollTop;
}
function hasScrollbar(target) {
    return target.scrollHeight > target.clientHeight;
}
var isOverflowing = function (container) {
    if (document.body === container) {
        return window.innerWidth > document.documentElement.clientWidth;
    }
    return container.scrollHeight > container.clientHeight;
};
var getPaddingRight = function (node) {
    return parseInt(window.getComputedStyle(node).paddingRight, 10) || 0;
};
var restoreContainerStyles = function (container) {
    var modalRestoreStyles = coreComponentsGlobalStore.getModalStore().getRestoreStyles();
    var index = modalRestoreStyles.findIndex(function (s) { return s.container === container; });
    var existingStyles = modalRestoreStyles[index];
    if (!existingStyles)
        return;
    existingStyles.modals -= 1;
    if (existingStyles.modals <= 0) {
        modalRestoreStyles.splice(index, 1);
        existingStyles.styles.forEach(function (_a) {
            var value = _a.value, el = _a.el, key = _a.key;
            if (value) {
                el.style.setProperty(key, value);
            }
            else {
                el.style.removeProperty(key);
            }
        });
    }
};
var handleContainer = function (container, shouldIOSLock) {
    if (shouldIOSLock === void 0) { shouldIOSLock = false; }
    if (!container)
        return;
    var modalRestoreStyles = coreComponentsGlobalStore.getModalStore().getRestoreStyles();
    var existingStyles = modalRestoreStyles.find(function (s) { return s.container === container; });
    if (existingStyles) {
        existingStyles.modals += 1;
        return;
    }
    var containerStyles = [];
    if (isOverflowing(container)) {
        // Вычисляет размер до применения `overflow hidden` для избежания скачков
        var scrollbarSize = coreComponentsShared.browser.getScrollbarSize();
        containerStyles.push({
            value: container.style.paddingRight,
            key: 'padding-right',
            el: container,
        });
        // Вычисляем стили, чтобы получить реальный `padding` c шириной сколлбара
        // eslint-disable-next-line no-param-reassign
        container.style.paddingRight = "".concat(getPaddingRight(container) + scrollbarSize, "px");
    }
    var parent = container.parentElement;
    var scrollContainer = 
    // TODO: заменить на optional chaining
    parent &&
        parent.nodeName === 'HTML' &&
        window.getComputedStyle(parent).overflowY === 'scroll'
        ? parent
        : container;
    // Блокируем скролл даже если отсутствует скроллбар
    if (scrollContainer.style.overflow !== 'hidden') {
        containerStyles.push({
            value: scrollContainer.style.overflow,
            key: 'overflow',
            el: scrollContainer,
        });
    }
    if (!shouldIOSLock) {
        scrollContainer.style.overflow = 'hidden';
    }
    modalRestoreStyles.push({
        container: container,
        modals: 1,
        styles: containerStyles,
    });
};

exports.handleContainer = handleContainer;
exports.hasScrollbar = hasScrollbar;
exports.isScrolledToBottom = isScrolledToBottom;
exports.isScrolledToTop = isScrolledToTop;
exports.restoreContainerStyles = restoreContainerStyles;
