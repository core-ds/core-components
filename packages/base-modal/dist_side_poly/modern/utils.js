import { getModalStore } from '@alfalab/core-components-global-store/modern';
import { browser } from '@alfalab/core-components-shared/modern';

function isScrolledToTop(target) {
    return target.scrollTop <= 0;
}
function isScrolledToBottom(target) {
    return target.scrollHeight - target.offsetHeight <= target.scrollTop;
}
function hasScrollbar(target) {
    return target.scrollHeight > target.clientHeight;
}
const isOverflowing = (container) => {
    if (document.body === container) {
        return window.innerWidth > document.documentElement.clientWidth;
    }
    return container.scrollHeight > container.clientHeight;
};
const getPaddingRight = (node) => parseInt(window.getComputedStyle(node).paddingRight, 10) || 0;
const restoreContainerStyles = (container) => {
    const modalRestoreStyles = getModalStore().getRestoreStyles();
    const index = modalRestoreStyles.findIndex((s) => s.container === container);
    const existingStyles = modalRestoreStyles[index];
    if (!existingStyles)
        return;
    existingStyles.modals -= 1;
    if (existingStyles.modals <= 0) {
        modalRestoreStyles.splice(index, 1);
        existingStyles.styles.forEach(({ value, el, key }) => {
            if (value) {
                el.style.setProperty(key, value);
            }
            else {
                el.style.removeProperty(key);
            }
        });
    }
};
const handleContainer = (container, shouldIOSLock = false) => {
    if (!container)
        return;
    const modalRestoreStyles = getModalStore().getRestoreStyles();
    const existingStyles = modalRestoreStyles.find((s) => s.container === container);
    if (existingStyles) {
        existingStyles.modals += 1;
        return;
    }
    const containerStyles = [];
    if (isOverflowing(container)) {
        // Вычисляет размер до применения `overflow hidden` для избежания скачков
        const scrollbarSize = browser.getScrollbarSize();
        containerStyles.push({
            value: container.style.paddingRight,
            key: 'padding-right',
            el: container,
        });
        // Вычисляем стили, чтобы получить реальный `padding` c шириной сколлбара
        // eslint-disable-next-line no-param-reassign
        container.style.paddingRight = `${getPaddingRight(container) + scrollbarSize}px`;
    }
    const parent = container.parentElement;
    const scrollContainer = 
    // TODO: заменить на optional chaining
    parent &&
        parent.nodeName === 'HTML' &&
        window.getComputedStyle(parent).overflowY === 'scroll'
        ? parent
        : container;
    // Блокируем скролл даже если отсутствует скроллбар
    if (scrollContainer.style.overflow !== 'hidden') {
        containerStyles.push({
            value: scrollContainer.style.overflow,
            key: 'overflow',
            el: scrollContainer,
        });
    }
    if (!shouldIOSLock) {
        scrollContainer.style.overflow = 'hidden';
    }
    modalRestoreStyles.push({
        container,
        modals: 1,
        styles: containerStyles,
    });
};

export { handleContainer, hasScrollbar, isScrolledToBottom, isScrolledToTop, restoreContainerStyles };
