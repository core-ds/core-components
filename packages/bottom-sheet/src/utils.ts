export const TIMEOUT = 500;
export const SWIPE_VELOCITY = 0.4;
export const MARKER_HEIGHT = 24;
/* Верхний отступ шторки, если она открыта на максимальную высоту */
export const HEADER_OFFSET = 24;
export const CLOSE_OFFSET = 0.2;
export const SCROLL_OFFSET = 1;

export const convertPercentToNumber = (value: string | number, fullHeight: number) => {
    const maxHeight = fullHeight - HEADER_OFFSET;

    if (typeof value === 'string') {
        const percent = parseFloat(value) / 100;

        if (percent < 0) {
            return Math.min(maxHeight, fullHeight + fullHeight * percent);
        }

        return Math.min(maxHeight, fullHeight * percent);
    }

    if (value < 0) {
        return Math.max(0, fullHeight + value);
    }

    return Math.min(maxHeight, value);
};
