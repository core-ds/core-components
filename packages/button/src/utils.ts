import { getDataTestId } from '@alfalab/core-components-shared';

export function getButtonTestIds(dataTestId: string) {
    return {
        button: dataTestId,
        spinner: getDataTestId(dataTestId, 'loader'),
    };
}
