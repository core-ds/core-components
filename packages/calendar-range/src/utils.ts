import { getDataTestId } from '@alfalab/core-components-shared';

export const isDayButton = (node: HTMLElement | null) =>
    node && node.tagName === 'BUTTON' && node.dataset.date;

export function getCalendarRangeTestIds(dataTestId: string) {
    return {
        component: dataTestId,
        calendarContainerFrom: getDataTestId(dataTestId, 'container-from'),
        calendarContainerTo: getDataTestId(dataTestId, 'container-to'),
    };
}
