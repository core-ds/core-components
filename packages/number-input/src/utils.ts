export const SIGNS = ['-', '+'];
export const SEPARATORS = [',', '.'];

export function createSeparatorsRegExp() {
    return new RegExp(`[${SEPARATORS.map((s) => `\\${s}`).join('')}]`, 'g');
}
