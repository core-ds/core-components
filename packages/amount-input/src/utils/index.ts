/**
 * https://github.com/core-ds/utils
 * TODO: нужно перенести утилиты в этот проект, когда он будет готов
 */

import { CurrencyCodes } from '@alfalab/data';
import { formatAmount } from '@alfalab/utils';

/**
 * Форматирует введенное значение
 * @param enteredValue Значение введенное в инпут
 * @param currency валюта
 * @param minority количество минорных единиц
 */
export function getFormattedValue(enteredValue: string, currency: CurrencyCodes, minority: number) {
    if (!enteredValue || enteredValue === '-') {
        return enteredValue;
    }

    // eslint-disable-next-line prefer-const
    let [head, tail] = enteredValue.split(',');

    // При вводе "-," указываем, что имеется в виду "-0,"
    if (head === '-') {
        head = '-0';
    }

    let { majorPart } = formatAmount({
        value: Number(head) * minority,
        currency,
        minority,
        negativeSymbol: 'hyphen-minus',
    });

    // Так как -0 === 0, formatAmount возвращает положительное значение. Исправляем это здесь
    if (/^-(,|0),?/.test(enteredValue) && majorPart === '0') {
        majorPart = `-${majorPart}`;
    }

    if (!tail && enteredValue.includes(',')) {
        return majorPart.concat(',');
    }

    if (tail) {
        return majorPart.concat(',', tail.slice(0, minority.toString().length - 1));
    }

    return majorPart;
}

export function getAmountValueFromStr(str: string, minority: number) {
    if (str === '' || str === '-') {
        return null;
    }

    return Math.round(Number(str.replace(',', '.').replace(/[^0-9.-]/g, '')) * minority);
}
