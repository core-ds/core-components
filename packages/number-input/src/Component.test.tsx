import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { getNumberInputTestIds } from './utils';

import { NumberInput, NumberInputProps } from '.';

describe('NumberInput', () => {
    Object.defineProperty(window, 'matchMedia', {
        writable: true,
        value: jest.fn().mockImplementation((query) => ({
            matches: true,
            media: query,
            onchange: null,
            addListener: jest.fn(), // Deprecated
            removeListener: jest.fn(), // Deprecated
            addEventListener: jest.fn(),
            removeEventListener: jest.fn(),
            dispatchEvent: jest.fn(),
        })),
    });

    describe('Snapshots tests', () => {
        it('should match snapshot', () => {
            expect(render(<NumberInput value='1234.567' />).container).toMatchSnapshot();
        });
    });

    it('should set `data-test-id` attribute', () => {
        const dti = 'number-input-dti';
        const { getByTestId } = render(
            <NumberInput
                block={true}
                dataTestId={dti}
                error='error message'
                leftAddons={<span />}
                rightAddons={<span />}
            />,
        );

        const testIds = getNumberInputTestIds(dti);

        expect(getByTestId(testIds.input)).toBeInTheDocument();
        expect(getByTestId(testIds.inputWrapper)).toBeInTheDocument();
        expect(getByTestId(testIds.inputWrapperInner)).toBeInTheDocument();
        expect(getByTestId(testIds.leftAddons)).toBeInTheDocument();
        expect(getByTestId(testIds.rightAddons)).toBeInTheDocument();
        expect(getByTestId(testIds.error)).toBeInTheDocument();
    });

    it('should set `data-test-id` attribute for hint, decrement, increment', () => {
        const dti = 'number-input-dti';
        const { getByTestId } = render(
            <NumberInput block={true} dataTestId={dti} step={1} hint='hint' />,
        );

        const testIds = getNumberInputTestIds(dti);

        expect(getByTestId(testIds.decrementButton)).toBeInTheDocument();
        expect(getByTestId(testIds.incrementButton)).toBeInTheDocument();
        expect(getByTestId(testIds.hint)).toBeInTheDocument();
    });

    it('should set custom class', () => {
        const className = 'custom-class';
        const { container } = render(<NumberInput className={className} />);

        expect(container.firstElementChild).toHaveClass(className);
    });

    it('should set `disabled` atribute', () => {
        const dataTestId = 'test-id';
        const incrementButtonId = dataTestId + '-increment-button';
        const decrementButtonId = dataTestId + '-decrement-button';
        const { getByTestId } = render(
            <NumberInput disabled={true} dataTestId={dataTestId} step={1} />,
        );

        expect(getByTestId(dataTestId)).toHaveAttribute('disabled');
        expect(getByTestId(incrementButtonId)).toHaveAttribute('disabled');
        expect(getByTestId(decrementButtonId)).toHaveAttribute('disabled');

        expect(getByTestId(dataTestId)).toHaveAttribute('disabled');
    });

    it('should decrement button disabled', () => {
        const dataTestId = 'test-id';
        const decrementButtonId = dataTestId + '-decrement-button';
        const incrementButtonId = dataTestId + '-increment-button';
        const { getByTestId } = render(
            <NumberInput value={10} min={10} dataTestId={dataTestId} step={1} />,
        );

        expect(getByTestId(decrementButtonId)).toHaveAttribute('disabled');
        expect(getByTestId(incrementButtonId)).not.toHaveAttribute('disabled');
    });

    it('should increment button disabled', () => {
        const dataTestId = 'test-id';
        const decrementButtonId = dataTestId + '-decrement-button';
        const incrementButtonId = dataTestId + '-increment-button';
        const { getByTestId } = render(
            <NumberInput value={10} max={10} dataTestId={dataTestId} step={1} />,
        );

        expect(getByTestId(decrementButtonId)).not.toHaveAttribute('disabled');
        expect(getByTestId(incrementButtonId)).toHaveAttribute('disabled');
    });

    it('should render error icon', () => {
        const { container } = render(<NumberInput error={true} />);

        expect(container.getElementsByClassName('errorIcon').length).toBe(1);
    });

    it('should render success icon', () => {
        const { container } = render(<NumberInput success={true} />);

        expect(container.getElementsByClassName('successIcon').length).toBe(1);
    });

    it('should not render success icon if has error', () => {
        const { container } = render(<NumberInput success={true} error={true} />);

        expect(container.getElementsByClassName('successIcon').length).toBe(0);
    });

    describe('Controlled-way', () => {
        it('should set value to input without minus sign', () => {
            const cb = jest.fn();
            const dataTestId = 'test-id';
            const { getByTestId } = render(
                <NumberInput onChange={cb} min={0} dataTestId={dataTestId} />,
            );

            const inputNumber = getByTestId(dataTestId) as HTMLInputElement;

            fireEvent.input(inputNumber, { target: { value: '-' } });
            expect(inputNumber.value).toBe('');

            fireEvent.input(inputNumber, { target: { value: '123' } });
            expect(inputNumber.value).toBe('123');
        });
    });

    describe('Callbacks tests', () => {
        it('should increase value by 3', () => {
            const cb = jest.fn();
            const dataTestId = 'test-id';
            const incrementButtonId = dataTestId + '-increment-button';
            const { getByTestId } = render(
                <NumberInput value={10} onChange={cb} dataTestId={dataTestId} step={3} />,
            );

            fireEvent.click(getByTestId(incrementButtonId));

            expect(cb).toBeCalledWith(expect.any(Object), { value: 13 });
        });

        it('should decrease value by 3', () => {
            const cb = jest.fn();
            const dataTestId = 'test-id';
            const decrementButtonId = dataTestId + '-decrement-button';
            const { getByTestId } = render(
                <NumberInput value={10} onChange={cb} dataTestId={dataTestId} step={3} />,
            );

            fireEvent.click(getByTestId(decrementButtonId));

            expect(cb).toBeCalledWith(expect.any(Object), { value: 7 });
        });

        it('should call onChange with valid value if max min is set', () => {
            const cb = jest.fn();
            const dataTestId = 'test-id';
            const { rerender, getByTestId } = render(
                <NumberInput min={10} max={50} onChange={cb} dataTestId={dataTestId} />,
            );

            fireEvent.input(getByTestId(dataTestId), { target: { value: 1 } });
            expect(cb).toBeCalledWith(expect.any(Object), { value: 1 });

            rerender(<NumberInput min={-50} max={-10} onChange={cb} dataTestId={dataTestId} />);

            fireEvent.input(getByTestId(dataTestId), { target: { value: -1 } });
            expect(cb).toBeCalledWith(expect.any(Object), { value: -1 });

            fireEvent.input(getByTestId(dataTestId), { target: { value: 1 } });
            expect(cb).toBeCalledWith(expect.any(Object), { value: -10 });
        });

        it('should call onChange with max value if value > max', () => {
            const cb = jest.fn();
            const dataTestId = 'test-id';
            const { getByTestId } = render(
                <NumberInput max={99} onChange={cb} dataTestId={dataTestId} />,
            );

            fireEvent.input(getByTestId(dataTestId), { target: { value: 101 } });
            expect(cb).toBeCalledWith(expect.any(Object), { value: 99 });
        });

        it('should call onChange with min value if value < min', () => {
            const cb = jest.fn();
            const dataTestId = 'test-id';
            const { getByTestId } = render(
                <NumberInput min={10} onChange={cb} dataTestId={dataTestId} />,
            );

            fireEvent.input(getByTestId(dataTestId), { target: { value: 5 } });
            expect(cb).toBeCalledWith(expect.any(Object), { value: 5 });
            fireEvent.blur(getByTestId(dataTestId));
            expect(cb).toBeCalledWith(expect.any(Object), { value: 10 });
        });

        it('should call `onChange` prop', () => {
            const cb = jest.fn();
            const dataTestId = 'test-id';
            const value = '123';
            const { getByTestId } = render(<NumberInput onChange={cb} dataTestId={dataTestId} />);

            const inputNumber = getByTestId(dataTestId) as HTMLInputElement;

            fireEvent.input(inputNumber, { target: { value } });

            expect(cb).toBeCalledTimes(1);
            expect(inputNumber.value).toBe(value);
        });

        it('should call `onFocus` prop', () => {
            const cb = jest.fn();
            const dataTestId = 'test-id';
            const { getByTestId } = render(<NumberInput onFocus={cb} dataTestId={dataTestId} />);

            fireEvent.focus(getByTestId(dataTestId));

            expect(cb).toBeCalledTimes(1);
        });

        it('should call `onBlur` prop', () => {
            const cb = jest.fn();
            const dataTestId = 'test-id';
            const { getByTestId } = render(<NumberInput onBlur={cb} dataTestId={dataTestId} />);

            fireEvent.blur(getByTestId(dataTestId));

            expect(cb).toBeCalledTimes(1);
        });

        it('should not call `onChange` prop if disabled', async () => {
            const cb = jest.fn();
            const dataTestId = 'test-id';
            const { getByTestId } = render(
                <NumberInput onChange={cb} dataTestId={dataTestId} disabled={true} />,
            );

            const inputNumber = getByTestId(dataTestId) as HTMLInputElement;

            await userEvent.type(inputNumber, '123');

            expect(cb).not.toBeCalled();
        });

        it('should call `onClear` prop when clear button clicked', () => {
            const cb = jest.fn();
            const { getByLabelText } = render(
                <NumberInput onClear={cb} clear={true} value='123' />,
            );

            fireEvent.click(getByLabelText('Очистить'));

            expect(cb).toBeCalledTimes(1);
        });

        it('should call `onChange` with valid value, when initial state is bad', async () => {
            const cb = jest.fn();
            const dataTestId = 'test-id';
            const { getByTestId } = render(
                <NumberInput onChange={cb} dataTestId={dataTestId} value='Вася123,Гена33' />,
            );

            const inputNumber = getByTestId(dataTestId) as HTMLInputElement;

            await userEvent.type(inputNumber, '{backspace}');

            expect(cb).toBeCalledWith(expect.any(Object), { value: 123.3 });
        });
    });

    describe('Render tests', () => {
        test('should unmount without errors', () => {
            const { unmount } = render(<NumberInput value='123' onChange={jest.fn()} />);

            expect(unmount).not.toThrowError();
        });
    });
});
