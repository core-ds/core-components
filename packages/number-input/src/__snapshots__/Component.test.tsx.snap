// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumberInput Snapshots tests should match snapshot 1`] = `
<div>
  <div
    class="component component s s"
  >
    <div
      class="inner inner inner inner filled"
    >
      <div
        class="inputWrapper"
      >
        <div
          class="input"
        >
          <input
            class="input input input"
            inputmode="decimal"
            type="text"
            value="1234.567"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
